Improve error message when :exc:`TypeError` occurs during :meth:`dict.update`
